zkevm-circuits/src/keccak_circuit/keccak_padding.rs

```diff
@@ -0,0 +1,427 @@
+        )?;
+        Ok(())
+    }
+}
+
+impl<F: Field> KeccakPaddingConfig<F> {
+    pub(crate) fn configure(meta: &mut ConstraintSystem<F>) -> Self {
+        let q_enable = meta.selector();
+        let q_end = meta.advice_column();
+        let d_bits = [(); KECCAK_WIDTH].map(|_| meta.advice_column());
+        let d_lens = [(); KECCAK_RATE_IN_BYTES].map(|_| meta.advice_column());
+        let d_rlcs = [(); KECCAK_RATE_IN_BYTES].map(|_| meta.advice_column());
+        let s_flags = [(); KECCAK_RATE_IN_BYTES].map(|_| meta.advice_column());
+        let randomness = meta.advice_column();
+
+        meta.create_gate("boolean checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            //TODO: could be removed if combined with keccak circuit?
+            for data_bit in d_bits {
+                let b = meta.query_advice(data_bit, Rotation::cur());
+                cb.require_boolean("input data bit", b);
+            }
+
+            for s_flag in s_flags {
+                let s = meta.query_advice(s_flag, Rotation::cur());
+                cb.require_boolean("boolean state bit", s);
+            }
+
+            for i in 1..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                let s_i_1 = meta.query_advice(s_flags[i - 1], Rotation::cur());
+
+                cb.require_boolean("boolean state switch", s_i - s_i_1);
+            }
+
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("padding bit checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            let mut constraints = vec![];
+            for i in 1..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                let s_i_1 = meta.query_advice(s_flags[i - 1], Rotation::cur());
+                let d_bit_0 = meta.query_advice(d_bits[8 * i], Rotation::cur());
+                constraints.push(("begin with 1", (s_i - s_i_1) * (d_bit_0 - 1u64.expr())));
+            }
+            let s_last = meta.query_advice(s_flags[s_flags.len() - 1], Rotation::cur());
+            let d_last = meta.query_advice(d_bits[KECCAK_RATE - 1], Rotation::cur());
+
+            constraints.push(("end with 1", s_last * (d_last - 1u64.expr())));
+
+            cb.add_constraints(constraints);
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("intermedium 0 checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            let mut sum_padding_bits = 0u64.expr();
+            for i in 0..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                sum_padding_bits = d_bits[i * 8..(i + 1) * 8]
+                    .iter()
+                    .map(|b| meta.query_advice(*b, Rotation::cur()))
+                    .fold(sum_padding_bits, |sum, b| sum + s_i.clone() * b);
+            }
```

Comment:
This will indeed work here (all data bits are enforced to be boolean so we don't have to worry about overflows for any practical data size) and will also be a bit more efficient! In the real keccak circuit the data bits will be split over multiple rows so you'll have to do this check taking that into account (but easy enough it seems).